pilot/pkg/xds/eds.go

```diff
@@ -165,9 +167,38 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
 	// For existing endpoints, we need to do full push if service accounts change.
 	if saUpdated {
 		log.Infof("Full push, service accounts changed, %v", hostname)
-		fullPush = true
+		return true, true
+	}
+	if created {
+		return true, true
+	}
+	if !features.SendUnhealthyEndpoints {
+		return true, true
 	}
-	return fullPush
+	// Check if new Endpoints are ready to be pushed. This check
+	// will ensure that if a new pod comes with a non ready endpoint,
+	// we do not unnecessarily push that config to Envoy.
+	emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+	for _, oie := range oldIstioEndpoints {
+		emap[oie.Address] = oie
```

Comment:
I would note (in the comments as well) that address is not a unique key. So this may push too many. Which is ok since its an optimization, but should be clearly noted.